lib/Dialect/LLHD/IR/LLHDCanonicalization.cpp

```diff
@@ -3,13 +3,53 @@
 //===----------------------------------------------------------------------===//
 
 #include "circt/Dialect/LLHD/IR/LLHDOps.h"
+#include "mlir/IR/Matchers.h"
 #include "mlir/IR/PatternMatch.h"
 
 using namespace mlir;
 
 namespace {
 /// Include the patterns defined in the Declarative Rewrite framework.
 #include "circt/Dialect/LLHD/IR/LLHDCanonicalization.inc"
+
+struct DynExtractSliceWithConstantStart
+    : public mlir::OpRewritePattern<llhd::DynExtractSliceOp> {
+  DynExtractSliceWithConstantStart(mlir::MLIRContext *context)
+      : OpRewritePattern<llhd::DynExtractSliceOp>(context, /*benefit=*/1) {}
+
+  mlir::LogicalResult
+  matchAndRewrite(llhd::DynExtractSliceOp op,
+                  mlir::PatternRewriter &rewriter) const override {
+    IntegerAttr intAttr;
+    if (mlir::matchPattern(op.start(), m_Constant<IntegerAttr>(&intAttr))) {
+      rewriter.replaceOpWithNewOp<llhd::ExtractSliceOp>(
+          op, op.result().getType(), op.target(),
+          rewriter.getIndexAttr(intAttr.getValue().getZExtValue()));
+      return success();
+    }
+    return failure();
+  }
+};
+
+struct DynExtractElementWithConstantIndex
+    : public mlir::OpRewritePattern<llhd::DynExtractElementOp> {
+  DynExtractElementWithConstantIndex(mlir::MLIRContext *context)
```

Comment:
I think these would be clearer if tablegen'd?

Reply:
I replaced them with tablegen patterns. Do you know a better way to match on `ConstantOp` and `LLHD_ConstOp` at the same time. Creating two patterns for what is actually only one pattern doesn't look that nice. Something like the `m_Constant<IntegerAttr>()` matcher in tablegen would be nice, but couldn't find an alternative so far.
This is necessary because the arithmetic operations we use from the standard dialect fold to `ConstantOp` and the LLHD operations to `LLHD_ConstOp`. Maybe it would be better to drop integer support from `LLHD_ConstOp` and only use `ConstantOp`...